lib/experimental/results.nim

```diff
@@ -0,0 +1,273 @@
+##[
+This module implements a result-type which encapsulates both a success and a
+failure in one value. Only one of the two states can be active.
+
+Based on https://github.com/disruptek/badresults, which was itself based on
+https://github.com/arnetheduck/nim-result.
+
+Exception translation
+=====================
+
+Using `get` or `value` raises an exception if the `Result` stores a failure.
+For `Result[T, E]` the exception raised is:
+ * if `E` is a `ref` type that inherits from `Exception`, the `error` value
+ * if a custom overload of `toException <#toException, E>`_ exists, the result
+  of `toException(r)`
+ * otherwise, a new `ResultError[E]` with `error` as the payload
+
+]##
+
+import std/options
+
+type
+  ResultError*[E] = ref object of ValueError
+    error: E
+
+  Result*[T, E] = object
+    ## A `Result` is typically used to augment a return value of type `T`
+    ## indicating that a computation (proc, etc...) is uncertain and should
+    ## return a value of type `T` or end in an error of type `E`. This is
+    ## useful to avoid encoding entirely expected error cases as values instead
+    ## of resorting to exception handling for program control flow or various
+    ## side-channel data.
+    case o: bool
+    of false:
+      e: E
+    of true:
+      v: T
+
+func error*[T, E](self: Result[T, E]): lent E {.inline.} =
+  ## Retrieves the error from a `Result`; raises `ResultError[void]`
+  ## if the `Result` is not an error.
+  if self.isErr:
+    result = self.e
+  else:
+    raise ResultError[void](msg: "Result does not contain an error")
+
+func error[T, E](self: var Result[T, E]): var E {.inline.} =
+  ## Retrieves the error from a `Result`; raises `ResultError[void]`
+  ## if the `Result` is not an error. For internal use only
+  if self.isErr:
+    result = self.e
+  else:
+    raise ResultError[void](msg: "Result does not contain an error")
```

Comment:
Is this used for anything at all? I think direct `.e` access would be better for internal code.